### PR TITLE
[FLINK-6334] [table] Refactoring UDTF interface

### PR DESCRIPTION
**Pull request by Xpray**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/StreamTableEnvironment.scala [hunk 1]
```diff
@@ -19,11 +19,12 @@ package org.apache.flink.table.api.java
 
 import org.apache.flink.api.common.typeinfo.TypeInformation
 import org.apache.flink.api.java.typeutils.TypeExtractor
-import org.apache.flink.table.api._
-import org.apache.flink.table.functions.TableFunction
-import org.apache.flink.table.expressions.ExpressionParser
 import org.apache.flink.streaming.api.datastream.DataStream
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
+import org.apache.flink.table.api._
+import org.apache.flink.table.api.scala.TableFunctionConversions
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/StreamTableEnvironment.scala]: Please remove the useless import.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/TableFunctionConversions.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.flink.table.functions.TableFunction
+import org.apache.flink.table.plan.logical.LogicalTableFunctionCall
+
+/**
+  * Holds methods to convert a [[TableFunction]] (provided by scala user) into a [[Table]]
+  *
+  * @param tf The tableFunction to convert.
+  */
+class TableFunctionConversions[T](tf: TableFunction[T]) {
+
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/TableFunctionConversions.scala]: I think before apply the  `TableFunction` ,It's just a definition. And when it's applied. It's a table. So I like named `AppliedTableFunction`. So we have two step a bout using `UDTF`. that is: define -> apply -> join.
What do you think?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 3]
```diff
@@ -417,12 +452,33 @@ class Table(
 
   private def join(right: Table, joinPredicate: Option[Expression], joinType: JoinType): Table = {
     // check that right table belongs to the same TableEnvironment
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: // check that  the TableEnvironment of right table is not null and right table belongs to the same TableEnvironment

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala [hunk 4]
```diff
@@ -20,8 +20,8 @@ package org.apache.flink.table.api.java
 import org.apache.flink.api.common.typeinfo.TypeInformation
 import org.apache.flink.api.java.typeutils.TypeExtractor
 import org.apache.flink.api.java.{DataSet, ExecutionEnvironment}
-import org.apache.flink.table.expressions.ExpressionParser
 import org.apache.flink.table.api._
+import org.apache.flink.table.expressions.ExpressionParser
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala]: Please remove this line.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala [hunk 5]
```diff
@@ -20,8 +20,8 @@ package org.apache.flink.table.api.java
 import org.apache.flink.api.common.typeinfo.TypeInformation
 import org.apache.flink.api.java.typeutils.TypeExtractor
 import org.apache.flink.api.java.{DataSet, ExecutionEnvironment}
-import org.apache.flink.table.expressions.ExpressionParser
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala]: Please restore this line.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/TableFunctionConversions.scala (hunk 2 above)
Comment (Xpray) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/TableFunctionConversions.scala]: I found that most of the existing implicit convert functions have common postfix like XXXConversions, to which I think this naming may be clear.

Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/TableFunctionConversions.scala]: In fact, I think it's a `Type` not a `Conversion`. I'am not sure. I think @fhueske can give us bestest suggestion.

**Comment by sunjincheng121** [comment]
HI @Xpray Thanks for the update. I think the PR is in a pretty good shape for me. Except for the name of `TableFunctionConversions`.  looking forward to @fhueske 's opinion.

Best,
SunJincheng

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/TableFunctionConversions.scala (hunk 2 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/TableFunctionConversions.scala]: I think the name `TableFunctionConversions` is OK. A `TableFunction` object is implicitly converted into a `TableFunctionConversions` and the `apply` method converts it into a `Table`.

**Pull request closed without merging**